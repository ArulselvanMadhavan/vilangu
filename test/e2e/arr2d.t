int main() {
  int a[][], i;
  int j;
  int col;
  a = new int[5][];
  i = 5;
  col = 4;
  while(i > 0){
    i = i - 1;
    j = col;
    a[i] = new int[j];
    while(j > 0){
      j = j - 1;
      a[i][j] = i * col + j;
      out a[i][j];      
    }
  }
}