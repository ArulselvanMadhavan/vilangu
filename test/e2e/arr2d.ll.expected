; ModuleID = 'Module'
source_filename = "Module"
target triple = "x86_64-pc-linux-gnu"

%Object_Vtable_type = type { %Object_Vtable_type*, i8* }
%i32arr_Vtable_type = type { %Object_Vtable_type*, i8* }
%i32arrarr_Vtable_type = type { %Object_Vtable_type*, i8* }
%Object = type { %Object_Vtable_type* }
%i32arr = type { %i32arr_Vtable_type*, i32*, i32 }
%i32arrarr = type { %i32arrarr_Vtable_type*, %i32arr*, i32 }

@printIntFormat = private constant [4 x i8] c"%d\0A\00", align 1
@castErrFormat = private constant [38 x i8] c"Narrow cast err: %s is not a type %s\0A\00", align 1
@outOfBoundsFormat = private constant [64 x i8] c"Array out of bounds exception at line %d. Index:%d | Length:%d\0A\00", align 1
@negativeLenFormat = private constant [51 x i8] c"Array len cannot be negative. Length attempted:%d\0A\00", align 1
@Object_class_name = private constant [7 x i8] c"Object\00", align 1
@Object_Vtable = global %Object_Vtable_type { %Object_Vtable_type* null, i8* getelementptr inbounds ([7 x i8], [7 x i8]* @Object_class_name, i32 0, i32 0) }
@i32arr_class_name = private constant [7 x i8] c"i32arr\00", align 1
@i32arr_Vtable = global %i32arr_Vtable_type { %Object_Vtable_type* @Object_Vtable, i8* getelementptr inbounds ([7 x i8], [7 x i8]* @i32arr_class_name, i32 0, i32 0) }
@i32arrarr_class_name = private constant [10 x i8] c"i32arrarr\00", align 1
@i32arrarr_Vtable = global %i32arrarr_Vtable_type { %Object_Vtable_type* @Object_Vtable, i8* getelementptr inbounds ([10 x i8], [10 x i8]* @i32arrarr_class_name, i32 0, i32 0) }

declare i32 @printf(i8*, ...)

declare i8* @malloc(i32)

declare i8* @calloc(i32, i32)

declare void @free(i8*)

declare void @exit(i32)

define i1 @Object_IsA(%Object* %0, %Object_Vtable_type* %1, i8* %2) {
entry:
  %object = alloca %Object*, align 8
  store %Object* %0, %Object** %object, align 8
  %vtbl = alloca %Object_Vtable_type*, align 8
  store %Object_Vtable_type* %1, %Object_Vtable_type** %vtbl, align 8
  %name = alloca i8*, align 8
  store i8* %2, i8** %name, align 8
  %3 = load %Object*, %Object** %object, align 8
  %4 = ptrtoint %Object* %3 to i64
  %equal = icmp eq i64 %4, 0
  %not = xor i1 %equal, true
  br i1 %not, label %then, label %else7

then:                                             ; preds = %entry
  %5 = load %Object*, %Object** %object, align 8
  %6 = getelementptr inbounds %Object, %Object* %5, i32 0, i32 0
  %7 = load %Object_Vtable_type*, %Object_Vtable_type** %6, align 8
  store %Object_Vtable_type* %7, %Object_Vtable_type** %vtbl, align 8
  br label %loopcond

loopcond:                                         ; preds = %ifcont, %then
  %8 = load %Object_Vtable_type*, %Object_Vtable_type** %vtbl, align 8
  %9 = ptrtoint %Object_Vtable_type* %8 to i64
  %equal1 = icmp eq i64 %9, 0
  %not2 = xor i1 %equal1, true
  br i1 %not2, label %loop, label %loopend

loop:                                             ; preds = %loopcond
  %10 = load %Object_Vtable_type*, %Object_Vtable_type** %vtbl, align 8
  %11 = getelementptr inbounds %Object_Vtable_type, %Object_Vtable_type* %10, i32 0, i32 1
  %12 = load i8*, i8** %11, align 8
  %13 = load i8*, i8** %name, align 8
  %14 = ptrtoint i8* %12 to i64
  %15 = ptrtoint i8* %13 to i64
  %equal3 = icmp eq i64 %14, %15
  br i1 %equal3, label %then4, label %else

then4:                                            ; preds = %loop
  br label %loopend

else:                                             ; preds = %loop
  %16 = load %Object_Vtable_type*, %Object_Vtable_type** %vtbl, align 8
  %17 = getelementptr inbounds %Object_Vtable_type, %Object_Vtable_type* %16, i32 0, i32 0
  %18 = load %Object_Vtable_type*, %Object_Vtable_type** %17, align 8
  store %Object_Vtable_type* %18, %Object_Vtable_type** %vtbl, align 8
  br label %ifcont

ifcont:                                           ; preds = %else
  br label %loopcond

loopend:                                          ; preds = %then4, %loopcond
  %19 = load %Object_Vtable_type*, %Object_Vtable_type** %vtbl, align 8
  %20 = ptrtoint %Object_Vtable_type* %19 to i64
  %equal5 = icmp eq i64 %20, 0
  %not6 = xor i1 %equal5, true
  br label %ifcont8

else7:                                            ; preds = %entry
  br label %ifcont8

ifcont8:                                          ; preds = %else7, %loopend
  %iftmp = phi i1 [ %not6, %loopend ], [ false, %else7 ]
  ret i1 %iftmp
}

define void @i32arr_Constructor(%i32arr* %0, %i32arr_Vtable_type* %1, i32* %2, i32 %3) {
entry:
  %this = alloca %i32arr*, align 8
  store %i32arr* %0, %i32arr** %this, align 8
  %vtable = alloca %i32arr_Vtable_type*, align 8
  store %i32arr_Vtable_type* %1, %i32arr_Vtable_type** %vtable, align 8
  %data = alloca i32*, align 8
  store i32* %2, i32** %data, align 8
  %length = alloca i32, align 4
  store i32 %3, i32* %length, align 4
  %4 = load %i32arr_Vtable_type*, %i32arr_Vtable_type** %vtable, align 8
  %5 = load %i32arr*, %i32arr** %this, align 8
  %6 = getelementptr inbounds %i32arr, %i32arr* %5, i32 0, i32 0
  store %i32arr_Vtable_type* %4, %i32arr_Vtable_type** %6, align 8
  %7 = load i32*, i32** %data, align 8
  %8 = load %i32arr*, %i32arr** %this, align 8
  %9 = getelementptr inbounds %i32arr, %i32arr* %8, i32 0, i32 1
  store i32* %7, i32** %9, align 8
  %10 = load i32, i32* %length, align 4
  %11 = load %i32arr*, %i32arr** %this, align 8
  %12 = getelementptr inbounds %i32arr, %i32arr* %11, i32 0, i32 2
  store i32 %10, i32* %12, align 4
  ret void
}

define void @i32arrarr_Constructor(%i32arrarr* %0, %i32arrarr_Vtable_type* %1, %i32arr* %2, i32 %3) {
entry:
  %this = alloca %i32arrarr*, align 8
  store %i32arrarr* %0, %i32arrarr** %this, align 8
  %vtable = alloca %i32arrarr_Vtable_type*, align 8
  store %i32arrarr_Vtable_type* %1, %i32arrarr_Vtable_type** %vtable, align 8
  %data = alloca %i32arr*, align 8
  store %i32arr* %2, %i32arr** %data, align 8
  %length = alloca i32, align 4
  store i32 %3, i32* %length, align 4
  %4 = load %i32arrarr_Vtable_type*, %i32arrarr_Vtable_type** %vtable, align 8
  %5 = load %i32arrarr*, %i32arrarr** %this, align 8
  %6 = getelementptr inbounds %i32arrarr, %i32arrarr* %5, i32 0, i32 0
  store %i32arrarr_Vtable_type* %4, %i32arrarr_Vtable_type** %6, align 8
  %7 = load %i32arr*, %i32arr** %data, align 8
  %8 = load %i32arrarr*, %i32arrarr** %this, align 8
  %9 = getelementptr inbounds %i32arrarr, %i32arrarr* %8, i32 0, i32 1
  store %i32arr* %7, %i32arr** %9, align 8
  %10 = load i32, i32* %length, align 4
  %11 = load %i32arrarr*, %i32arrarr** %this, align 8
  %12 = getelementptr inbounds %i32arrarr, %i32arrarr* %11, i32 0, i32 2
  store i32 %10, i32* %12, align 4
  ret void
}

define i32 @main() {
entry:
  %col = alloca i32, align 4
  %j = alloca i32, align 4
  %i = alloca i32, align 4
  %a = alloca %i32arrarr*, align 8
  store %i32arrarr* null, %i32arrarr** %a, align 8
  store i32 0, i32* %i, align 4
  store i32 0, i32* %j, align 4
  store i32 0, i32* %col, align 4
  br i1 false, label %negativeLenThen1, label %negativeLenCont

negativeLenThen1:                                 ; preds = %entry
  br label %negativeLenCont

negativeLenCont:                                  ; preds = %negativeLenThen1, %entry
  %newArrayResult = alloca %i32arrarr, align 8
  %0 = call dereferenceable_or_null(120) i8* @calloc(i32 5, i32 24)
  %1 = bitcast i8* %0 to %i32arr*
  call void @i32arrarr_Constructor(%i32arrarr* nonnull %newArrayResult, %i32arrarr_Vtable_type* nonnull @i32arrarr_Vtable, %i32arr* %1, i32 5)
  store %i32arrarr* %newArrayResult, %i32arrarr** %a, align 8
  store i32 5, i32* %i, align 4
  store i32 4, i32* %col, align 4
  br label %loopcond

loopcond:                                         ; preds = %loopend, %negativeLenCont
  %2 = load i32, i32* %i, align 4
  %sgt = icmp sgt i32 %2, 0
  br i1 %sgt, label %loop, label %loopend34

loop:                                             ; preds = %loopcond
  %3 = load i32, i32* %i, align 4
  %sub = add i32 %3, -1
  store i32 %sub, i32* %i, align 4
  %4 = load i32, i32* %col, align 4
  store i32 %4, i32* %j, align 4
  %5 = icmp slt i32 %4, 0
  br i1 %5, label %negativeLenThen2, label %negativeLenCont3

negativeLenThen2:                                 ; preds = %loop
  %6 = call i32 (i8*, ...) @printf(i8* noundef nonnull dereferenceable(1) getelementptr inbounds ([51 x i8], [51 x i8]* @negativeLenFormat, i64 0, i64 0), i32 %4)
  call void @exit(i32 -1)
  br label %negativeLenCont3

negativeLenCont3:                                 ; preds = %negativeLenThen2, %loop
  %newArrayResult4 = alloca %i32arr, align 8
  %7 = call i8* @calloc(i32 %4, i32 4)
  %8 = bitcast i8* %7 to i32*
  call void @i32arr_Constructor(%i32arr* nonnull %newArrayResult4, %i32arr_Vtable_type* nonnull @i32arr_Vtable, i32* %8, i32 %4)
  %9 = load i32, i32* %i, align 4
  %10 = load %i32arrarr*, %i32arrarr** %a, align 8
  %11 = getelementptr inbounds %i32arrarr, %i32arrarr* %10, i64 0, i32 2
  %12 = load i32, i32* %11, align 4
  %13 = icmp slt i32 %9, %12
  br i1 %13, label %boundsCheckThen5, label %boundsCheckElse

boundsCheckThen5:                                 ; preds = %negativeLenCont3
  %14 = load %i32arrarr*, %i32arrarr** %a, align 8
  %15 = getelementptr inbounds %i32arrarr, %i32arrarr* %14, i64 0, i32 1
  %baseArrPtr = load %i32arr*, %i32arr** %15, align 8
  %16 = sext i32 %9 to i64
  %17 = getelementptr inbounds %i32arr, %i32arr* %baseArrPtr, i64 %16
  br label %boundsCheckIfcont

boundsCheckElse:                                  ; preds = %negativeLenCont3
  %18 = call i32 (i8*, ...) @printf(i8* noundef nonnull dereferenceable(1) getelementptr inbounds ([64 x i8], [64 x i8]* @outOfBoundsFormat, i64 0, i64 0), i32 11, i32 %9, i32 %12)
  call void @exit(i32 -1)
  br label %boundsCheckIfcont

boundsCheckIfcont:                                ; preds = %boundsCheckElse, %boundsCheckThen5
  %loadRefTypeBeforeSave = load %i32arr, %i32arr* %newArrayResult4, align 8
  store %i32arr %loadRefTypeBeforeSave, %i32arr* %17, align 8
  br label %loopcond6

loopcond6:                                        ; preds = %boundsCheckIfcont33, %boundsCheckIfcont
  %19 = load i32, i32* %j, align 4
  %sgt7 = icmp sgt i32 %19, 0
  br i1 %sgt7, label %loop8, label %loopend

loop8:                                            ; preds = %loopcond6
  %20 = load i32, i32* %j, align 4
  %sub9 = add i32 %20, -1
  store i32 %sub9, i32* %j, align 4
  %21 = load i32, i32* %i, align 4
  %22 = load i32, i32* %col, align 4
  %mul = mul i32 %21, %22
  %add = add i32 %mul, %sub9
  %23 = load %i32arrarr*, %i32arrarr** %a, align 8
  %24 = getelementptr inbounds %i32arrarr, %i32arrarr* %23, i64 0, i32 2
  %25 = load i32, i32* %24, align 4
  %26 = icmp slt i32 %21, %25
  br i1 %26, label %boundsCheckThen11, label %boundsCheckElse12

boundsCheckThen11:                                ; preds = %loop8
  %27 = load %i32arrarr*, %i32arrarr** %a, align 8
  %28 = getelementptr inbounds %i32arrarr, %i32arrarr* %27, i64 0, i32 1
  %baseArrPtr10 = load %i32arr*, %i32arr** %28, align 8
  %29 = sext i32 %21 to i64
  br label %boundsCheckIfcont13

boundsCheckElse12:                                ; preds = %loop8
  %30 = call i32 (i8*, ...) @printf(i8* noundef nonnull dereferenceable(1) getelementptr inbounds ([64 x i8], [64 x i8]* @outOfBoundsFormat, i64 0, i64 0), i32 14, i32 %21, i32 %25)
  call void @exit(i32 -1)
  br label %boundsCheckIfcont13

boundsCheckIfcont13:                              ; preds = %boundsCheckElse12, %boundsCheckThen11
  %31 = getelementptr inbounds %i32arr, %i32arr* %baseArrPtr10, i64 %29, i32 1
  %baseArrPtr14 = load i32*, i32** %31, align 8
  %32 = load i32, i32* %i, align 4
  %33 = load %i32arrarr*, %i32arrarr** %a, align 8
  %34 = getelementptr inbounds %i32arrarr, %i32arrarr* %33, i64 0, i32 2
  %35 = load i32, i32* %34, align 4
  %36 = icmp slt i32 %32, %35
  br i1 %36, label %boundsCheckThen16, label %boundsCheckElse17

boundsCheckThen16:                                ; preds = %boundsCheckIfcont13
  %37 = load %i32arrarr*, %i32arrarr** %a, align 8
  %38 = getelementptr inbounds %i32arrarr, %i32arrarr* %37, i64 0, i32 1
  %baseArrPtr15 = load %i32arr*, %i32arr** %38, align 8
  %39 = sext i32 %32 to i64
  br label %boundsCheckIfcont18

boundsCheckElse17:                                ; preds = %boundsCheckIfcont13
  %40 = call i32 (i8*, ...) @printf(i8* noundef nonnull dereferenceable(1) getelementptr inbounds ([64 x i8], [64 x i8]* @outOfBoundsFormat, i64 0, i64 0), i32 14, i32 %32, i32 %35)
  call void @exit(i32 -1)
  br label %boundsCheckIfcont18

boundsCheckIfcont18:                              ; preds = %boundsCheckElse17, %boundsCheckThen16
  %41 = getelementptr inbounds %i32arr, %i32arr* %baseArrPtr15, i64 %39, i32 2
  %42 = load i32, i32* %41, align 4
  %43 = icmp slt i32 %sub9, %42
  br i1 %43, label %boundsCheckThen19, label %boundsCheckElse20

boundsCheckThen19:                                ; preds = %boundsCheckIfcont18
  %44 = sext i32 %sub9 to i64
  %45 = getelementptr inbounds i32, i32* %baseArrPtr14, i64 %44
  br label %boundsCheckIfcont21

boundsCheckElse20:                                ; preds = %boundsCheckIfcont18
  %46 = call i32 (i8*, ...) @printf(i8* noundef nonnull dereferenceable(1) getelementptr inbounds ([64 x i8], [64 x i8]* @outOfBoundsFormat, i64 0, i64 0), i32 14, i32 %sub9, i32 %42)
  call void @exit(i32 -1)
  br label %boundsCheckIfcont21

boundsCheckIfcont21:                              ; preds = %boundsCheckElse20, %boundsCheckThen19
  store i32 %add, i32* %45, align 4
  %47 = load i32, i32* %j, align 4
  %48 = load i32, i32* %i, align 4
  %49 = load %i32arrarr*, %i32arrarr** %a, align 8
  %50 = getelementptr inbounds %i32arrarr, %i32arrarr* %49, i64 0, i32 2
  %51 = load i32, i32* %50, align 4
  %52 = icmp slt i32 %48, %51
  br i1 %52, label %boundsCheckThen23, label %boundsCheckElse24

boundsCheckThen23:                                ; preds = %boundsCheckIfcont21
  %53 = load %i32arrarr*, %i32arrarr** %a, align 8
  %54 = getelementptr inbounds %i32arrarr, %i32arrarr* %53, i64 0, i32 1
  %baseArrPtr22 = load %i32arr*, %i32arr** %54, align 8
  %55 = sext i32 %48 to i64
  br label %boundsCheckIfcont25

boundsCheckElse24:                                ; preds = %boundsCheckIfcont21
  %56 = call i32 (i8*, ...) @printf(i8* noundef nonnull dereferenceable(1) getelementptr inbounds ([64 x i8], [64 x i8]* @outOfBoundsFormat, i64 0, i64 0), i32 15, i32 %48, i32 %51)
  call void @exit(i32 -1)
  br label %boundsCheckIfcont25

boundsCheckIfcont25:                              ; preds = %boundsCheckElse24, %boundsCheckThen23
  %57 = getelementptr inbounds %i32arr, %i32arr* %baseArrPtr22, i64 %55, i32 1
  %baseArrPtr26 = load i32*, i32** %57, align 8
  %58 = load i32, i32* %i, align 4
  %59 = load %i32arrarr*, %i32arrarr** %a, align 8
  %60 = getelementptr inbounds %i32arrarr, %i32arrarr* %59, i64 0, i32 2
  %61 = load i32, i32* %60, align 4
  %62 = icmp slt i32 %58, %61
  br i1 %62, label %boundsCheckThen28, label %boundsCheckElse29

boundsCheckThen28:                                ; preds = %boundsCheckIfcont25
  %63 = load %i32arrarr*, %i32arrarr** %a, align 8
  %64 = getelementptr inbounds %i32arrarr, %i32arrarr* %63, i64 0, i32 1
  %baseArrPtr27 = load %i32arr*, %i32arr** %64, align 8
  %65 = sext i32 %58 to i64
  br label %boundsCheckIfcont30

boundsCheckElse29:                                ; preds = %boundsCheckIfcont25
  %66 = call i32 (i8*, ...) @printf(i8* noundef nonnull dereferenceable(1) getelementptr inbounds ([64 x i8], [64 x i8]* @outOfBoundsFormat, i64 0, i64 0), i32 15, i32 %58, i32 %61)
  call void @exit(i32 -1)
  br label %boundsCheckIfcont30

boundsCheckIfcont30:                              ; preds = %boundsCheckElse29, %boundsCheckThen28
  %67 = getelementptr inbounds %i32arr, %i32arr* %baseArrPtr27, i64 %65, i32 2
  %68 = load i32, i32* %67, align 4
  %69 = icmp slt i32 %47, %68
  br i1 %69, label %boundsCheckThen31, label %boundsCheckElse32

boundsCheckThen31:                                ; preds = %boundsCheckIfcont30
  %70 = sext i32 %47 to i64
  %71 = getelementptr inbounds i32, i32* %baseArrPtr26, i64 %70
  br label %boundsCheckIfcont33

boundsCheckElse32:                                ; preds = %boundsCheckIfcont30
  %72 = call i32 (i8*, ...) @printf(i8* noundef nonnull dereferenceable(1) getelementptr inbounds ([64 x i8], [64 x i8]* @outOfBoundsFormat, i64 0, i64 0), i32 15, i32 %47, i32 %68)
  call void @exit(i32 -1)
  br label %boundsCheckIfcont33

boundsCheckIfcont33:                              ; preds = %boundsCheckElse32, %boundsCheckThen31
  %73 = load i32, i32* %71, align 4
  %74 = call i32 (i8*, ...) @printf(i8* noundef nonnull dereferenceable(1) getelementptr inbounds ([4 x i8], [4 x i8]* @printIntFormat, i64 0, i64 0), i32 %73)
  br label %loopcond6

loopend:                                          ; preds = %loopcond6
  br label %loopcond

loopend34:                                        ; preds = %loopcond
  ret i32 0
}
